File: fixtures/go/calculator_test.go
package calculator

import "testing"

func TestAdd(t *testing.T) {
	tests := []struct {
		name     string
		a, b     int
		expected int
	}{
		{"positive numbers", 2, 3, 5},
		{"negative numbers", -1, -1, -2},
		{"zero", 0, 5, 5},
		{"mixed", -3, 5, 2},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Add(tt.a, tt.b)
			if result != tt.expected {
				t.Errorf("Add(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
			}
		})
	}
}

func TestSubtract(t *testing.T) {
	tests := []struct {
		a, b     int
		expected int
	}{
		{5, 3, 2},
		{10, 5, 5},
		{0, 5, -5},
		{-3, -3, 0},
	}

	for _, tt := range tests {
		if result := Subtract(tt.a, tt.b); result != tt.expected {
			t.Errorf("Subtract(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestMultiply(t *testing.T) {
	tests := []struct {
		a, b     int
		expected int
	}{
		{4, 5, 20},
		{0, 10, 0},
		{-2, 3, -6},
		{-2, -3, 6},
	}

	for _, tt := range tests {
		if result := Multiply(tt.a, tt.b); result != tt.expected {
			t.Errorf("Multiply(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestDivide(t *testing.T) {
	tests := []struct {
		a, b     int
		expected int
	}{
		{10, 2, 5},
		{20, 4, 5},
		{15, 3, 5},
		{-10, 2, -5},
	}

	for _, tt := range tests {
		if result := Divide(tt.a, tt.b); result != tt.expected {
			t.Errorf("Divide(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestModulo(t *testing.T) {
	tests := []struct {
		a, b     int
		expected int
	}{
		{10, 3, 1},
		{15, 4, 3},
		{20, 5, 0},
		{7, 3, 1},
	}

	for _, tt := range tests {
		if result := Modulo(tt.a, tt.b); result != tt.expected {
			t.Errorf("Modulo(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestIsPositive(t *testing.T) {
	tests := []struct {
		value    int
		expected bool
	}{
		{5, true},
		{0, false},
		{-5, false},
		{100, true},
		{-1, false},
	}

	for _, tt := range tests {
		if result := IsPositive(tt.value); result != tt.expected {
			t.Errorf("IsPositive(%d) = %v; want %v", tt.value, result, tt.expected)
		}
	}
}

func TestIsGreaterOrEqual(t *testing.T) {
	tests := []struct {
		a, b     int
		expected bool
	}{
		{5, 3, true},
		{3, 3, true},
		{3, 5, false},
		{0, 0, true},
		{-1, 1, false},
	}

	for _, tt := range tests {
		if result := IsGreaterOrEqual(tt.a, tt.b); result != tt.expected {
			t.Errorf("IsGreaterOrEqual(%d, %d) = %v; want %v", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestIsEqual(t *testing.T) {
	tests := []struct {
		a, b     int
		expected bool
	}{
		{5, 5, true},
		{5, 3, false},
		{0, 0, true},
		{-1, -1, true},
		{-1, 1, false},
	}

	for _, tt := range tests {
		if result := IsEqual(tt.a, tt.b); result != tt.expected {
			t.Errorf("IsEqual(%d, %d) = %v; want %v", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestBothPositive(t *testing.T) {
	tests := []struct {
		a, b     int
		expected bool
	}{
		{5, 10, true},
		{5, -1, false},
		{-1, 5, false},
		{-1, -1, false},
		{0, 5, false},
		{5, 0, false},
	}

	for _, tt := range tests {
		if result := BothPositive(tt.a, tt.b); result != tt.expected {
			t.Errorf("BothPositive(%d, %d) = %v; want %v", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestEitherPositive(t *testing.T) {
	tests := []struct {
		a, b     int
		expected bool
	}{
		{5, 10, true},
		{5, -1, true},
		{-1, 5, true},
		{-1, -1, false},
		{0, 0, false},
	}

	for _, tt := range tests {
		if result := EitherPositive(tt.a, tt.b); result != tt.expected {
			t.Errorf("EitherPositive(%d, %d) = %v; want %v", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestBitwiseAnd(t *testing.T) {
	tests := []struct {
		a, b     int
		expected int
	}{
		{6, 3, 2},   // 110 & 011 = 010
		{12, 10, 8}, // 1100 & 1010 = 1000
		{15, 7, 7},  // 1111 & 0111 = 0111
	}

	for _, tt := range tests {
		if result := BitwiseAnd(tt.a, tt.b); result != tt.expected {
			t.Errorf("BitwiseAnd(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestBitwiseOr(t *testing.T) {
	tests := []struct {
		a, b     int
		expected int
	}{
		{6, 3, 7},   // 110 | 011 = 111
		{12, 10, 14}, // 1100 | 1010 = 1110
		{8, 4, 12},  // 1000 | 0100 = 1100
	}

	for _, tt := range tests {
		if result := BitwiseOr(tt.a, tt.b); result != tt.expected {
			t.Errorf("BitwiseOr(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestBitwiseXor(t *testing.T) {
	tests := []struct {
		a, b     int
		expected int
	}{
		{6, 3, 5},   // 110 ^ 011 = 101
		{12, 10, 6}, // 1100 ^ 1010 = 0110
		{15, 15, 0}, // Same values = 0
	}

	for _, tt := range tests {
		if result := BitwiseXor(tt.a, tt.b); result != tt.expected {
			t.Errorf("BitwiseXor(%d, %d) = %d; want %d", tt.a, tt.b, result, tt.expected)
		}
	}
}

func TestLeftShift(t *testing.T) {
	tests := []struct {
		a, shift int
		expected int
	}{
		{1, 3, 8},  // 1 << 3 = 8
		{5, 2, 20}, // 5 << 2 = 20
		{3, 1, 6},  // 3 << 1 = 6
	}

	for _, tt := range tests {
		if result := LeftShift(tt.a, tt.shift); result != tt.expected {
			t.Errorf("LeftShift(%d, %d) = %d; want %d", tt.a, tt.shift, result, tt.expected)
		}
	}
}

func TestRightShift(t *testing.T) {
	tests := []struct {
		a, shift int
		expected int
	}{
		{8, 3, 1},  // 8 >> 3 = 1
		{20, 2, 5}, // 20 >> 2 = 5
		{6, 1, 3},  // 6 >> 1 = 3
	}

	for _, tt := range tests {
		if result := RightShift(tt.a, tt.shift); result != tt.expected {
			t.Errorf("RightShift(%d, %d) = %d; want %d", tt.a, tt.shift, result, tt.expected)
		}
	}
}

func TestNegate(t *testing.T) {
	tests := []struct {
		value    int
		expected int
	}{
		{5, -5},
		{-5, 5},
		{0, 0},
		{100, -100},
	}

	for _, tt := range tests {
		if result := Negate(tt.value); result != tt.expected {
			t.Errorf("Negate(%d) = %d; want %d", tt.value, result, tt.expected)
		}
	}
}

func TestPositive(t *testing.T) {
	tests := []struct {
		value    int
		expected int
	}{
		{5, 5},
		{-5, -5}, // Unary + doesn't change sign
		{0, 0},
	}

	for _, tt := range tests {
		if result := Positive(tt.value); result != tt.expected {
			t.Errorf("Positive(%d) = %d; want %d", tt.value, result, tt.expected)
		}
	}
}

func TestNot(t *testing.T) {
	tests := []struct {
		flag     bool
		expected bool
	}{
		{true, false},
		{false, true},
	}

	for _, tt := range tests {
		if result := Not(tt.flag); result != tt.expected {
			t.Errorf("Not(%v) = %v; want %v", tt.flag, result, tt.expected)
		}
	}
}

func TestAddAssign(t *testing.T) {
	tests := []struct {
		value, delta int
		expected     int
	}{
		{10, 5, 15},
		{0, 10, 10},
		{-5, 3, -2},
	}

	for _, tt := range tests {
		if result := AddAssign(tt.value, tt.delta); result != tt.expected {
			t.Errorf("AddAssign(%d, %d) = %d; want %d", tt.value, tt.delta, result, tt.expected)
		}
	}
}

func TestSubtractAssign(t *testing.T) {
	tests := []struct {
		value, delta int
		expected     int
	}{
		{10, 5, 5},
		{0, 10, -10},
		{-5, 3, -8},
	}

	for _, tt := range tests {
		if result := SubtractAssign(tt.value, tt.delta); result != tt.expected {
			t.Errorf("SubtractAssign(%d, %d) = %d; want %d", tt.value, tt.delta, result, tt.expected)
		}
	}
}

func TestMultiplyAssign(t *testing.T) {
	tests := []struct {
		value, factor int
		expected      int
	}{
		{10, 5, 50},
		{3, 4, 12},
		{-2, 3, -6},
	}

	for _, tt := range tests {
		if result := MultiplyAssign(tt.value, tt.factor); result != tt.expected {
			t.Errorf("MultiplyAssign(%d, %d) = %d; want %d", tt.value, tt.factor, result, tt.expected)
		}
	}
}

func TestDivideAssign(t *testing.T) {
	tests := []struct {
		value, divisor int
		expected       int
	}{
		{20, 5, 4},
		{15, 3, 5},
		{-10, 2, -5},
	}

	for _, tt := range tests {
		if result := DivideAssign(tt.value, tt.divisor); result != tt.expected {
			t.Errorf("DivideAssign(%d, %d) = %d; want %d", tt.value, tt.divisor, result, tt.expected)
		}
	}
}

File: fixtures/rust/tests/calculator_test.rs
// tests/calculator_test.rs
// Comprehensive test suite for Rust calculator
// Demonstrates mutation testing with various operators

use rust_fixtures::calculator::*;

// ============================================================
// ARITHMETIC OPERATOR TESTS (AOR)
// ============================================================

#[test]
fn test_add() {
    assert_eq!(add(2, 3), 5);
    assert_eq!(add(-1, -1), -2);
    assert_eq!(add(0, 5), 5);
    assert_eq!(add(-3, 5), 2);
    assert_eq!(add(100, 200), 300);
}

#[test]
fn test_subtract() {
    assert_eq!(subtract(5, 3), 2);
    assert_eq!(subtract(-1, -1), 0);
    assert_eq!(subtract(0, 5), -5);
    assert_eq!(subtract(-3, 5), -8);
    assert_eq!(subtract(10, 3), 7);
}

#[test]
fn test_multiply() {
    assert_eq!(multiply(2, 3), 6);
    assert_eq!(multiply(-2, 3), -6);
    assert_eq!(multiply(0, 5), 0);
    assert_eq!(multiply(-2, -3), 6);
    assert_eq!(multiply(7, 8), 56);
}

#[test]
fn test_divide() {
    assert_eq!(divide(6, 3), 2);
    assert_eq!(divide(-6, 3), -2);
    assert_eq!(divide(5, 2), 2); // Integer division
    assert_eq!(divide(5, 0), 0); // Division by zero
    assert_eq!(divide(20, 4), 5);
}

#[test]
fn test_modulo() {
    assert_eq!(modulo(7, 3), 1);
    assert_eq!(modulo(10, 5), 0);
    assert_eq!(modulo(5, 2), 1);
    assert_eq!(modulo(5, 0), 0); // Modulo by zero
    assert_eq!(modulo(13, 4), 1);
}

// ============================================================
// RELATIONAL OPERATOR TESTS (ROR)
// ============================================================

#[test]
fn test_greater_than() {
    assert!(greater_than(5, 3));
    assert!(!greater_than(3, 5));
    assert!(!greater_than(3, 3));
    assert!(greater_than(100, 50));
    assert!(!greater_than(-5, 0));
}

#[test]
fn test_less_than() {
    assert!(less_than(3, 5));
    assert!(!less_than(5, 3));
    assert!(!less_than(3, 3));
    assert!(less_than(-10, 0));
    assert!(!less_than(100, 50));
}

#[test]
fn test_greater_or_equal() {
    assert!(greater_or_equal(5, 3));
    assert!(greater_or_equal(3, 3));
    assert!(!greater_or_equal(3, 5));
    assert!(greater_or_equal(10, 10));
    assert!(!greater_or_equal(5, 10));
}

#[test]
fn test_less_or_equal() {
    assert!(less_or_equal(3, 5));
    assert!(less_or_equal(3, 3));
    assert!(!less_or_equal(5, 3));
    assert!(less_or_equal(10, 10));
    assert!(!less_or_equal(15, 10));
}

#[test]
fn test_equal() {
    assert!(equal(3, 3));
    assert!(!equal(3, 5));
    assert!(equal(0, 0));
    assert!(equal(-5, -5));
    assert!(!equal(10, 11));
}

#[test]
fn test_not_equal() {
    assert!(not_equal(3, 5));
    assert!(!not_equal(3, 3));
    assert!(not_equal(10, 20));
    assert!(!not_equal(-5, -5));
    assert!(not_equal(0, 1));
}

// ============================================================
// LOGICAL OPERATOR TESTS (LOR)
// ============================================================

#[test]
fn test_logical_and() {
    assert!(logical_and(true, true));
    assert!(!logical_and(true, false));
    assert!(!logical_and(false, true));
    assert!(!logical_and(false, false));
}

#[test]
fn test_logical_or() {
    assert!(logical_or(true, true));
    assert!(logical_or(true, false));
    assert!(logical_or(false, true));
    assert!(!logical_or(false, false));
}

// ============================================================
// BITWISE OPERATOR TESTS (BOR)
// ============================================================

#[test]
fn test_bitwise_and() {
    assert_eq!(bitwise_and(5, 3), 1);   // 101 & 011 = 001
    assert_eq!(bitwise_and(12, 10), 8); // 1100 & 1010 = 1000
    assert_eq!(bitwise_and(15, 7), 7);  // 1111 & 0111 = 0111
    assert_eq!(bitwise_and(0, 255), 0);
}

#[test]
fn test_bitwise_or() {
    assert_eq!(bitwise_or(5, 3), 7);    // 101 | 011 = 111
    assert_eq!(bitwise_or(12, 10), 14); // 1100 | 1010 = 1110
    assert_eq!(bitwise_or(8, 4), 12);   // 1000 | 0100 = 1100
    assert_eq!(bitwise_or(0, 255), 255);
}

#[test]
fn test_bitwise_xor() {
    assert_eq!(bitwise_xor(5, 3), 6);   // 101 ^ 011 = 110
    assert_eq!(bitwise_xor(12, 10), 6); // 1100 ^ 1010 = 0110
    assert_eq!(bitwise_xor(15, 15), 0); // Same values = 0
    assert_eq!(bitwise_xor(255, 0), 255);
}

#[test]
fn test_left_shift() {
    assert_eq!(left_shift(5, 1), 10);  // 101 << 1 = 1010
    assert_eq!(left_shift(3, 2), 12);  // 011 << 2 = 1100
    assert_eq!(left_shift(1, 4), 16);  // 1 << 4 = 10000
    assert_eq!(left_shift(7, 3), 56);
}

#[test]
fn test_right_shift() {
    assert_eq!(right_shift(10, 1), 5); // 1010 >> 1 = 101
    assert_eq!(right_shift(12, 2), 3); // 1100 >> 2 = 011
    assert_eq!(right_shift(16, 4), 1); // 10000 >> 4 = 1
    assert_eq!(right_shift(100, 2), 25);
}

#[test]
fn test_bitwise_not() {
    assert_eq!(bitwise_not(5), !5);
    assert_eq!(bitwise_not(0), -1);
    assert_eq!(bitwise_not(-1), 0);
    assert_eq!(bitwise_not(255), !255);
}

// ============================================================
// RANGE OPERATOR TESTS (RANGEOR) - RUST-SPECIFIC
// ============================================================

#[test]
fn test_exclusive_range_sum() {
    assert_eq!(exclusive_range_sum(0, 5), 10);  // 0+1+2+3+4
    assert_eq!(exclusive_range_sum(1, 4), 6);   // 1+2+3
    assert_eq!(exclusive_range_sum(5, 5), 0);   // Empty range
}

#[test]
fn test_inclusive_range_sum() {
    assert_eq!(inclusive_range_sum(0, 5), 15);  // 0+1+2+3+4+5
    assert_eq!(inclusive_range_sum(1, 4), 10);  // 1+2+3+4
    assert_eq!(inclusive_range_sum(5, 5), 5);   // Single element
}

#[test]
fn test_exclusive_range_collect() {
    assert_eq!(exclusive_range_collect(0, 5), vec![0, 1, 2, 3, 4]);
    assert_eq!(exclusive_range_collect(1, 4), vec![1, 2, 3]);
    assert_eq!(exclusive_range_collect(5, 5), vec![] as Vec<i32>);
}

#[test]
fn test_inclusive_range_collect() {
    assert_eq!(inclusive_range_collect(0, 5), vec![0, 1, 2, 3, 4, 5]);
    assert_eq!(inclusive_range_collect(1, 4), vec![1, 2, 3, 4]);
    assert_eq!(inclusive_range_collect(5, 5), vec![5]);
}

// ============================================================
// PATTERN MATCHING TESTS (PMR) - RUST-SPECIFIC
// ============================================================

#[test]
fn test_unwrap_option() {
    assert_eq!(unwrap_option(Some(42)), 42);
    assert_eq!(unwrap_option(None), 0);
    assert_eq!(unwrap_option(Some(-5)), -5);
    assert_eq!(unwrap_option(Some(0)), 0);
}

#[test]
fn test_unwrap_result() {
    assert_eq!(unwrap_result(Ok(42)), 42);
    assert_eq!(unwrap_result(Err("error".to_string())), 0);
    assert_eq!(unwrap_result(Ok(-5)), -5);
    assert_eq!(unwrap_result(Ok(0)), 0);
}

#[test]
fn test_is_some() {
    assert!(is_some(Some(42)));
    assert!(!is_some(None));
    assert!(is_some(Some(0)));
}

#[test]
fn test_is_ok() {
    assert!(is_ok(Ok(42)));
    assert!(!is_ok(Err("error".to_string())));
    assert!(is_ok(Ok(0)));
}

// ============================================================
// METHOD CHAINING TESTS (MCR) - RUST-SPECIFIC
// ============================================================

#[test]
fn test_map_filter_example() {
    let input = vec![1, 2, 3, 4, 5];
    let result = map_filter_example(input);
    assert_eq!(result, vec![6, 8, 10]);  // [2,4,6,8,10] filtered to >5
}

#[test]
fn test_filter_map_example() {
    let input = vec![-2, -1, 0, 1, 2, 3];
    let result = filter_map_example(input);
    assert_eq!(result, vec![2, 4, 6]);  // Positive values doubled
}

#[test]
fn test_unwrap_or_example() {
    assert_eq!(unwrap_or_example(Some(42)), 42);
    assert_eq!(unwrap_or_example(None), 0);
}

#[test]
fn test_unwrap_or_default_example() {
    assert_eq!(unwrap_or_default_example(Some(42)), 42);
    assert_eq!(unwrap_or_default_example(None), 0);
}

// ============================================================
// BORROW/REFERENCE TESTS (LBM) - RUST-SPECIFIC
// ============================================================

#[test]
fn test_borrow_immutable() {
    let value = 42;
    assert_eq!(borrow_immutable(&value), 42);

    let value2 = -5;
    assert_eq!(borrow_immutable(&value2), -5);
}

#[test]
fn test_borrow_mutable() {
    let mut value = 42;
    borrow_mutable(&mut value);
    assert_eq!(value, 43);

    borrow_mutable(&mut value);
    assert_eq!(value, 44);
}

#[test]
fn test_return_reference() {
    let value = 42;
    let borrowed = return_reference(&value);
    assert_eq!(*borrowed, 42);
}

#[test]
fn test_return_mutable_reference() {
    let mut value = 42;
    let borrowed = return_mutable_reference(&mut value);
    assert_eq!(*borrowed, 43);

    *borrowed += 1;
    assert_eq!(value, 44);
}
